### PR TITLE
Add an 'interface switcher' to the toolbar

### PR DESCRIPTION
**Pull request by yuvipanda**
Fixes https://github.com/jupyterlab/retrolab/issues/157

image

Is how it looks. Can switch to lab or classic.

### CODE REVIEW

**Comment by yuvipanda** [comment]
I also wonder if we should trigger a save of the notebook first? Otherwise you might end up not seeing the proper notebook

**Comment by jtpio** [comment]
Thanks!

We still depend on `nbclassic` at the moment:

https://github.com/jupyterlab/retrolab/blob/0f7c026472e64b9f9073cce47a00c072d6a84af5/setup.py#L90

So there could indeed be an interface switcher like this in the notebook toolbar.

> I also wonder if we should trigger a save of the notebook first? Otherwise you might end up not seeing the proper notebook

Sounds good :+1:

**Comment by jtpio** [comment]
Also I'm wondering whether this plugin could actually go in the lab extension package? 

https://github.com/jupyterlab/retrolab/blob/main/packages/lab-extension/src/index.ts

Then it would also be available in the JupyterLab UI:

image

We could then remove the existing RetroLab icon from the notebook toolbar, since this would already be covered by the switch?

image

**Comment by yuvipanda** [comment]
@jtpio so if we move it to the labextension, we'll have to determine current setup (lab or retro) and adjust choices accordingly. That seems reasonable, I'll do that instead.

**Comment by jtpio** [comment]
We could yes, by checking whether there is an `IRetroShell` or `ILabShell` (by having them in `optional` in the plugin).

It would also be fine to have them all show up in both retro and lab to start with.

**Comment by yuvipanda** [comment]
@jtpio I can't seem to figure out how to test this in lab though :(

**Comment by yuvipanda** [comment]
I did refactor the code to suck less, and theoretically added code to behave differently when running in lab vs retro. Any idea how I can test it?

File: packages/lab-extension/src/interfaceswitcher.tsx [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+        dropdownLabel: 'Classic',
+        urlPrefix: `${baseUrl}tree/`,
+        current: false
+      },
+      {
+        command: 'retrolab:open-retro',
+        commandLabel: 'Open in RetroLab',
+        dropdownLabel: 'RetroLab',
+        urlPrefix: `${baseUrl}retro/tree/`,
+        current: app.name === 'RetroLab'
```
Comment (jtpio) [inline packages/lab-extension/src/interfaceswitcher.tsx:118]: Maybe we could also add `ILabShell` and `IRetroShell` as `optional` dependencies for the plugin, and check whether they are `null` to know if the current interface is a JupyterLab-like or RetroLab-like interface?

Comment (jtpio) [reply packages/lab-extension/src/interfaceswitcher.tsx:118]: `name` could indeed be overriden via the page config or directly in JS like here: 

https://github.com/jtpio/jupyterlite/blob/a198d902908483f8eea36fecae858cbd14b4a078/app/lab/index.template.js#L176

**Comment by jtpio** [comment]
Thanks Yuvi for the update :+1: 

Since the `lab-extension` package is a prebuild extension for JupyerLab, it should normally be loaded automatically in both interfaces.

For real testing, it would be great to start using end-to-end tools like Galata at some point: #135

File: app/index.js [hunk 2]
```diff
@@ -87,6 +87,9 @@ async function main() {
     require('@retrolab/docmanager-extension'),
     require('@retrolab/help-extension'),
     require('@retrolab/notebook-extension'),
+    require('@retrolab/lab-extension').default.filter(({ id }) =>
+      ['@retrolab/lab-extension:interface-switcher'].includes(id)
+    ),
```
Comment (jtpio) [inline app/index.js:92]:
```suggestion
```

**Comment by yuvipanda** [comment]
Ah, I'll make these changes :) And to clarify, when I said 'I do not know how to test these' I was refering to not being able to get JupyterLab to reflect my changes at all! So I can't actually see if what I did worked...

**Comment by jtpio** [comment]
It seems to be showing fine in both retro and lab now when testing with the Binder link above:

image

image

Although choosing one of the dropdown options doesn't seem to perform the switch.

Need to check, but maybe you need to run `python -m pip install -e .` every time for the changes to the lab extension to take effect locally. If so then we should definitely improve that workflow and use the `jupyter labextension develop` command.

**Comment by yuvipanda** [comment]
@jtpio yay, running `pip install -e .` worked. The idea is that instead of saying 'Interface' it will show the current interface by default, and selecting another will switch to that.

I'll change the way to detect which UI we are running by using `ILabShell` instead of the `.name`

**Comment by yuvipanda** [comment]
@jtpio gentle bump - anything else needed here?

**Comment by jtpio** [comment]
Thanks Yuvi for the update.

Checking with the Binder link above, it looks like choosing other interfaces has no effect?

https://user-images.githubusercontent.com/591645/122888268-fc0d5800-d341-11eb-810b-1692a2c10355.mp4

And it seems to be defaulting to JupyterLab.

**Comment by jtpio** [comment]
But using the commands from the command palette works fine.

**Comment by yuvipanda** [comment]
oh interesting - it works for me. I'll investigate it this week.

**Comment by jtpio** [comment]
@yuvipanda I pushed an update in https://github.com/jupyterlab/retrolab/pull/158/commits/3f49eea5519282f289df00b590a6514b56bd6ff0 to show the appropriate notebook toolbar buttons depending on the current interface:

https://user-images.githubusercontent.com/591645/131989639-9b77ef40-618f-4acf-8344-0291014f9a01.mp4

Using buttons instead of a dropdown to help simplify the state management.

If you want to try it on Binder:

https://mybinder.org/v2/gh/yuvipanda/retrolab/open-in?urlpath=retro/tree

If it looks good to you we can merge and ship it right after in `0.3.4`.

**Comment by yuvipanda** [comment]
Awesome! This looks great to me, thank you for putting effort into it!!!

**Comment by jtpio** [comment]
Great, thanks for checking it out!

Let's get it in then, I'll cut `0.3.4` right after.

**Comment by jtpio** [comment]
https://pypi.org/project/retrolab/0.3.4/

**Pull request merged**